### PR TITLE
DE/DA method refactor

### PR DESCRIPTION
**Pull request by justjhong**
This PR prepares the DE and DA functions for public use:
- Rename `perform_multivariate_analysis` to `differential_expression`
- Change `donor` to `sample` in many places
- Add `differential_abundance` method that wraps the `get_aggregated_posterior` method and uses a similar API to `differential_expression`
- Rewrites `test_models.py` to use pytest parameterize, and adds tests for DA.

Addresses #94 #95

### CODE REVIEW

File: src/mrvi/_model.py [hunk 1]
```diff
@@ -743,6 +753,111 @@ def get_aggregated_posterior(
             dist.Normal(qu_loc, qu_scale),
         )
 
+    def differential_abundance(
+        self,
+        adata: AnnData | None = None,
+        sample_cov_keys: list[str] | None = None,
+        sample_subset: list[str] | None = None,
+        minibatch_size: int = 128,
```
Comment (martinkim0) [inline src/mrvi/_model.py]:
```suggestion
        batch_size: int = 128,
```

File: src/mrvi/_model.py [hunk 2]
```diff
@@ -743,6 +753,111 @@ def get_aggregated_posterior(
...
+        ----------
+        adata
+            The data object to compute the differential abundance for.
+            If not given, the data object stored in the model is used.
+        sample_cov_keys
+            Keys for covariates (batch, etc.) that should also be taken into account
+            when computing the differential abundance.
+        sample_subset
+            Only compute differential abundance for these sample labels.
+        minibatch_size
```
Comment (martinkim0) [inline src/mrvi/_model.py]:
```suggestion
        batch_size
```

File: src/mrvi/_model.py [hunk 3]
```diff
@@ -743,6 +753,111 @@ def get_aggregated_posterior(
...
+
+        Returns
+        -------
+        :class:`xarray.Dataset`
+            Returns an xarray.Dataset with data variables:
+            - `log_probs`: Array of shape (n_cells, n_samples) containing the log probabilities for each cell across samples.
+            - `{cov_key}_log_probs`: For each key in `sample_cov_keys`, an array of shape (n_cells, n_cov_values) containing the log probabilities for each cell across covariate values.
+        """
+        adata = self._validate_anndata(adata)
+        us = self.get_latent_representation(adata, use_mean=True, give_z=False)
```
Comment (martinkim0) [inline src/mrvi/_model.py]: Pass in batch size here?

File: src/mrvi/_model.py [hunk 4]
```diff
@@ -743,6 +753,111 @@ def get_aggregated_posterior(
...
+
+        log_probs = []
+        unique_samples = adata.obs[self.sample_key].unique()
+        for sample_name in tqdm(unique_samples):
+            ap = self.get_aggregated_posterior(adata=adata, sample=sample_name)
```
Comment (martinkim0) [inline src/mrvi/_model.py]: Pass in batch size here too or nah?

File: src/mrvi/_model.py [hunk 5]
```diff
@@ -743,6 +753,111 @@ def get_aggregated_posterior(
...
+            n_splits = adata.n_obs // minibatch_size
+            log_probs_ = []
+            for u_rep in np.array_split(us, n_splits):
```
Comment (martinkim0) [inline src/mrvi/_model.py:836]: Will this error if `n_obs < minibatch_size`?

File: src/mrvi/_model.py [hunk 6]
```diff
@@ -743,6 +753,111 @@ def get_aggregated_posterior(
...
+                log_probs_.append(
+                    jax.device_get(ap.log_prob(u_rep).sum(-1, keepdims=True))
+                )
```
Comment (martinkim0) [inline src/mrvi/_model.py:839]: Could you clarify why this inner loop is necessary? Is `us` too large to do it all at once?

File: src/mrvi/_model.py [hunk 7]
```diff
@@ -878,22 +990,41 @@ def perform_multivariate_analysis(
...
         store_baseline
             Whether to store the expression in the module if logfoldchanges are computed.
         eps_lfc
             Epsilon to add to the log-fold changes to avoid detecting genes with low expression.
-        filter_donors
-            Whether to filter out-of-distribution donors prior to performing the analysis.
-        filter_donors_kwargs
+        filter_inadmissible_samples
+            Whether to filter out-of-distribution samples prior to performing the analysis.
+        filter_inadmissible_samples_kwargs
```
Comment (martinkim0) [inline src/mrvi/_model.py]:
```suggestion
        filter_samples_kwargs
```
Should be the last arg in the docstring

File: src/mrvi/_model.py [hunk 8]
```diff
@@ -878,22 +990,41 @@ def perform_multivariate_analysis(
...
+            - `effect_size`: Effect sizes for each covariate across cells.
+            - `pvalue`: P-values for each covariate across cells.
+            - `padj`: Adjusted P-values for each covariate across cells using the Benjamini-Hochberg procedure.
+            - `lfc`: Log fold changes for each covariate across cells and genes, if `store_lfc` is True.
+            - `lfc_std`: Standard deviation of log fold changes, if `store_lfc` is True and `delta` is not None.
+            - `pde`: Posterior DE probabilities, if `store_lfc` is True and `delta` is not None.
+            - `baseline_expression`: Baseline expression levels for each covariate across cells and genes, if `store_baseline` is True.
+            - `n_samples`: Number of admissible samples for each cell, if `filter_inadmissible_samples` is True.
         """
+        assert (
```
Comment (martinkim0) [inline src/mrvi/_model.py]: nit: use `ValueError`

File: tests/test_model.py [hunk 9]
```diff
@@ -1,297 +1,235 @@
...
+                },
+                {
+                    "sample_cov_keys": ["meta1_cat", "meta2", "cont_cov"],
+                    "store_lfc": True,
+                    "filter_inadmissible_samples": True,
+                },
+            ],
+        ),
+    ],
+)
```
Comment (martinkim0) [inline tests/test_model.py:99]: This is lowkey kind of hard to parse, is there a way to simplify it?

Comment (martinkim0) [reply tests/test_model.py:99]: ok nvm I see what it's doing now

File: src/mrvi/_model.py [hunk 10]
```diff
@@ -61,7 +62,7 @@ class MrVI(JaxTrainingMixin, BaseModelClass):
         AnnData object that has been registered via ``setup_anndata``.
     n_latent
         Dimensionality of the latent space.
-    n_latent_donor
+    n_latent_sample
```
Comment (martinkim0) [inline src/mrvi/_model.py]: This arg is being passed into `MrVAE` but it doesn't look like `MrVAE` has this argument defined.

Comment (justjhong) [reply src/mrvi/_model.py]: ah good catch, its an old one

File: tests/test_model.py (hunk 9 above)
Comment (justjhong) [reply tests/test_model.py:99]: yea..it's because its replicating the old behavior, which was not neatly combinatorial way, but I did not want to significantly change what was being tested. We can refactor it again if needed and it will be easier this time with parametrize

File: src/mrvi/_model.py (hunk 6 above)
Comment (justjhong) [reply src/mrvi/_model.py:839]: yes its splitting by samples, and samples have an arbitrary number of cells in them. To keep the range of one batch of computation consistent, I batch here

Comment (justjhong) [reply src/mrvi/_model.py:839]: And u can be big if one sample has a ton of cells for whatever reason

File: src/mrvi/_model.py (hunk 3 above)
Comment (justjhong) [reply src/mrvi/_model.py]: sure

File: src/mrvi/_model.py (hunk 4 above)
Comment (justjhong) [reply src/mrvi/_model.py]: yes sure, I think it makes sense to keep batches consistent if it controls the memory load

File: src/mrvi/_model.py (hunk 5 above)
Comment (justjhong) [reply src/mrvi/_model.py:836]: yes, good catch

File: src/mrvi/_model.py (hunk 6 above)
Comment (martinkim0) [reply src/mrvi/_model.py:839]: got it, makes sense

File: src/mrvi/_model.py [hunk 11]
```diff
@@ -743,13 +771,122 @@ def get_aggregated_posterior(
...
+                    continue
+
+                sel_log_probs = log_probs_arr.log_probs.loc[{"sample": cov_samples}]
+                val_log_probs = logsumexp(sel_log_probs, axis=1) - np.log(
+                    sel_log_probs.shape[1]
+                )
+                per_val_log_probs[sample_cov_value] = val_log_probs
+            sample_cov_log_probs_map[sample_cov_key] = pd.DataFrame.from_dict(
+                per_val_log_probs
+            )
```
Comment (PierreBoyeau) [inline src/mrvi/_model.py]: We should catch cases where sample_cov_key is continuous

Comment (PierreBoyeau) [reply src/mrvi/_model.py]: An option could be to raise a ValueError whenever the covariate has more than some n_covs_limit value

File: src/mrvi/_model.py [hunk 12]
```diff
@@ -743,13 +771,122 @@ def get_aggregated_posterior(
...
+        conditioned on the estimated aggregate posterior distribution of each cell.
+
+        Parameters
+        ----------
+        adata
+            The data object to compute the differential abundance for.
+            If not given, the data object stored in the model is used.
+        sample_cov_keys
+            Keys for covariates (batch, etc.) that should also be taken into account
+            when computing the differential abundance.
```
Comment (PierreBoyeau) [inline src/mrvi/_model.py]: specify that these are expected to be discrete. We could also, alternatively, have an internal procedure than bins continuous covariates. What would you prefer?

File: src/mrvi/_model.py [hunk 13]
```diff
@@ -743,13 +771,122 @@ def get_aggregated_posterior(
...
+
     def get_outlier_cell_sample_pairs(
         self,
         adata=None,
         subsample_size: int = 5_000,
         quantile_threshold: float = 0.05,
         admissibility_threshold: float = 0.0,
-        minibatch_size: int = 256,
+        batch_size: int = 256,
     ) -> xr.Dataset:
```
Comment (PierreBoyeau) [inline src/mrvi/_model.py:942]: add to docstring

File: src/mrvi/_model.py [hunk 14]
```diff
@@ -743,13 +771,122 @@ def get_aggregated_posterior(
...
+        batch_size
+            Minibatch size for computing the differential abundance.
+
+        Returns
+        -------
+        :class:`xarray.Dataset`
+            Returns an xarray.Dataset with data variables:
+            - `log_probs`: Array of shape (n_cells, n_samples) containing the log probabilities for each cell across samples.
+            - `{cov_key}_log_probs`: For each key in `sample_cov_keys`, an array of shape (n_cells, n_cov_values) containing the log probabilities for each cell across covariate values.
+        """
```
Comment (PierreBoyeau) [inline src/mrvi/_model.py:809]: an alternative or a complement to this output could also be to return the log enrichment scores, of the form
log_prob[:, current_cov_value] - log_prob[:, np.arange(log_prob.shape[1]) != current_cov_value].mean(-1)
I feel like many people could also be interested in this output.

**Review by PierreBoyeau — APPROVED** [review]
Looks good to me. I have some comments on differential_abudance, where I feel we could also return log enrichment scores and be more careful regarding continuous covarariates.

File: src/mrvi/_model.py (hunk 12 above)
Comment (justjhong) [reply src/mrvi/_model.py]: for now just expect discrete

File: src/mrvi/_model.py (hunk 11 above)
Comment (justjhong) [reply src/mrvi/_model.py]: throwing a warning for this
